# Guided Project: Exploring Hacker News Posts

## Introduction

In this project we will work with a data set of submissions to Hacker News.
Hacker News is a site made by the startup incubator Y Combinator, where posts are voted and commented upon. Hacker News is popular in technology circles, and posts that make it to the top of the website can get hundreds of thousands of visitors.

The data set can be found here: https://www.kaggle.com/hacker-news/hacker-news-posts

It had been reduced from almost 300,000 rows to approximately 20,000 rows by removing the submissions that don't have any comments, and them randomly sampling from the remaining submissions.

The description of the columns are as follow:

* id: The unique identifier from Hacker News for the post

* title: The title of the post

* url: The URL that the posts links to, if it the post has a URL

* num_points: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes

* num_comments: The number of comments that were made on the post

* author: The username of the person who submitted the post

* created_at: The date and time at which the post was submitted

We will be analyzing two types of posts:
 * Ask HN: Users submit this type of post when they want to ask questions to the community.
 * Show HN: Users submit this type of post when they want to show something they made or just something interesting in general

We'll compare them to answer two questions:

* Do Ask HN or Show HN receive more comments on average?

* Do posts created at a certain time receive more comments on average?

First we will import the libraries we'll need and read the data set:

In [18]:
from csv import reader

opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)

print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


We need to separate the header from the rest of the data set:

In [6]:
headers = hn[:1]

In [7]:
hn = hn[1:]

In [8]:
print(hn[:5])

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


Now that we've removed the headers from hn, we need to filter the post titles starting with "Ask HN" or "Show HN":

In [12]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(row)
    elif title.lower().startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)

Checking the number of posts for each newly created list:

In [14]:
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


Now let's see if ask posts or show posts receive more comments on average:

In [16]:
total_ask_comments = 0
for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments

avg_ask_comments = total_ask_comments/len(ask_posts)
print(avg_ask_comments)

total_show_comments = 0
for row in show_posts:
    num_comments = int(row[4])
    total_show_comments += num_comments

avg_show_comments = total_show_comments/len(show_posts)
print(avg_show_comments)


14.038417431192661
10.31669535283993


We now know that ask posts receive in average almost 4 more comments than show posts

Since ask posts receive more comments, we'll focus our analysis  on just these posts from now on.

Next we'll check if posts created at a certain time are more likely to attract comments.

In [20]:
import datetime as dt

result_list = []
for row in ask_posts:
    created = row[6]
    num_comments = int(row[4])
    result_list.append([created, num_comments])

counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for row in result_list:
    date = row[0]
    comments = row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time not in counts_by_hour:
        counts_by_hour[time] = 1
        comments_by_hour[time] = comments
    else:
        counts_by_hour[time] += 1
        comments_by_hour[time] += comments
        
comments_by_hour

{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

Now we created two dictionaries:
* counts_by_hour: contains the number of ask posts created during each hour of the day
* comments_by_hour: contains the corresponding number of comments ask posts created at each hour received.

We'll now calculate the average comments per post by hour:

In [22]:
avg_by_hour = []

for key in counts_by_hour:
    avg_by_hour.append([key, comments_by_hour[key]/counts_by_hour[key]])

avg_by_hour

[['22', 6.746478873239437],
 ['12', 9.41095890410959],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['21', 16.009174311926607],
 ['00', 8.127272727272727],
 ['05', 10.08695652173913],
 ['07', 7.852941176470588],
 ['06', 9.022727272727273],
 ['11', 11.051724137931034],
 ['03', 7.796296296296297],
 ['20', 21.525],
 ['15', 38.5948275862069],
 ['02', 23.810344827586206],
 ['19', 10.8],
 ['09', 5.5777777777777775],
 ['23', 7.985294117647059],
 ['16', 16.796296296296298],
 ['13', 14.741176470588234],
 ['14', 13.233644859813085],
 ['18', 13.20183486238532],
 ['01', 11.383333333333333],
 ['17', 11.46],
 ['10', 13.440677966101696]]

Sorting the table:

In [29]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])

sorted_swap = sorted(swap_avg_by_hour, reverse = True)

print("Top 5 Hours for Ask Post Comments")

for avg,hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for Ask Post Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The time zone is Eastern Time in the US: If we convert that to Brasília time we get:

16:00: 38.59 average comments per post

03:00: 23.81 average comments per post

21:00: 21.52 average comments per post

17:00: 16.80 average comments per post

22:00: 16.01 average comments per post

## Conclusion

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (4:00 pm Brasília - 5:00 pm Brasília).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (4:00 pm Brasília - 5:00 pm Brasília) received the most comments on average.